### ISSUE TITLE
PathMatchingResourcePatternResolver  gives me a FileSystemResource instead of ClassPathResource while I specify the prefix "classpath:" 

### ISSUE THREAD

**Issue by justmehyp** [spring-projects/spring-framework#25623]
Recently, I had some trouble loading classpath resources. During local development, I can obtain the resources under classpath
through classloader. And If this resource is a folder, I can get sub-files/ sub-directories through the `File#listfiles` method.
But this doesn't work when the program is packaged in a jar package, the file in a jar is not a file, and the folder is not a folder.
In this case, they are all `JarEntry`.

So I found out the PathMatchingResourcePatternResolver, which I hoped would solve my problem, but it couldn't.

The problem is described as follows:

First, we create a new PathMatchingResourcePatternResolver instance:
```
package com.xxx;

public class Main {
  public static void main(String[] args) {
    ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver();
  }
}
```

I want to get the resources under classpath, such as all sub-resources under 'org/springframework/core/io/support/',
```
resolver.getResources("classpath:org/springframework/core/io/support/*")
```

I specified the prefix "classpath:" just as I wanted to get resources under classpath.
The statement returns an Array with component type `ClassPathResource`. That is exactly what I want. And the class name
`ClassPathResource` is just so NICE.

OK, We just tried to get the resources inside a jar.
Next, We are going to get something in expanded directory(file system). My program has a package called` com.xxx `. Next, let's try to obtain the resources under "com/xxx":
```
resolver.getResources("classpath:com/xxx");
```

This time, take a closer look. It returns an Array with concrete component type `FileSystemResource`. I am confused now.
I specified the prefix "classpath:", instead of "file:///", so why give me a FileSystemResource?

I am expecting the perfect `ClassPathResource` because I need its `getPath` method to do other things.

Finally, I am thinking about that this may be a defect in implementation, and should be resolve. What do you guys think?

**Comment by Martin-Wegner** [comment]
Since 5.3.0 my package scanner is broken, which uses the ResourcePatternResolver to find classpath resources inside a JAR. I think this is related to this issue.

**Comment by Martin-Wegner** [comment]
@rstoyanchev why do you closed this issue? The bug is still present in Spring 5.3.13 :(

**Comment by rstoyanchev** [comment]
@Martin-Wegner it is closed as superseded by PR#25624. If both an issue and a PR are opened, we keep only one open. They are linked and there is no need to have both.

**Comment by Martin-Wegner** [comment]
@rstoyanchev thanks for this explanation :)
